zenith_utils/src/zid.rs

```diff
@@ -4,168 +4,119 @@ use hex::FromHex;
 use rand::Rng;
 use serde::{Deserialize, Serialize};
 
-// Zenith ID is a 128-bit random ID.
+// Zenith ID is a n-bit random ID.
 // Used to represent various identifiers. Provides handy utility methods and impls.
-#[derive(Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, PartialOrd, Ord)]
-struct ZId([u8; 16]);
-
-impl ZId {
-    pub fn get_from_buf(buf: &mut dyn bytes::Buf) -> ZId {
-        let mut arr = [0u8; 16];
-        buf.copy_to_slice(&mut arr);
-        ZId::from(arr)
-    }
-
-    pub fn as_arr(&self) -> [u8; 16] {
-        self.0
-    }
-
-    pub fn generate() -> Self {
-        let mut tli_buf = [0u8; 16];
-        rand::thread_rng().fill(&mut tli_buf);
-        ZId::from(tli_buf)
-    }
-}
-
-impl FromStr for ZId {
-    type Err = hex::FromHexError;
-
-    fn from_str(s: &str) -> Result<ZId, Self::Err> {
-        Self::from_hex(s)
-    }
-}
-
-// this is needed for pretty serialization and deserialization of ZId's using serde integration with hex crate
-impl FromHex for ZId {
-    type Error = hex::FromHexError;
-
-    fn from_hex<T: AsRef<[u8]>>(hex: T) -> Result<Self, Self::Error> {
-        let mut buf: [u8; 16] = [0u8; 16];
-        hex::decode_to_slice(hex, &mut buf)?;
-        Ok(ZId(buf))
-    }
-}
-
-impl AsRef<[u8]> for ZId {
-    fn as_ref(&self) -> &[u8] {
-        &self.0
-    }
-}
-
-impl From<[u8; 16]> for ZId {
-    fn from(b: [u8; 16]) -> Self {
-        ZId(b)
-    }
-}
-
-impl fmt::Display for ZId {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.write_str(&hex::encode(self.0))
-    }
-}
-
-impl fmt::Debug for ZId {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.write_str(&hex::encode(self.0))
-    }
-}
-
 macro_rules! zid_newtype {
-    ($t:ident) => {
-        impl $t {
-            pub fn get_from_buf(buf: &mut dyn bytes::Buf) -> $t {
-                $t(ZId::get_from_buf(buf))
+    // $attr meta argument is for attaching the doc string (which is in fact an
```

Comment:
I think that using a macro here is fine, since there's no support for const-generic-based arrays in serde, and fixing this with a boilerplate doesn't look much better (if at all). Instead, I'd propose to make ZNodeIds 128-bit to make this change unnecessary. Sure, that'd be somewhat more wasteful, but does it matter?

Reply:
Interesting idea, though I also think macro here is better. However I just decided to switch to sequential int8:
https://github.com/zenithdb/zenith/pull/1257#issuecomment-1035265127

The refactoring of zid.rs still makes sense to me though, removing basically empty wrapper type, so I'm going to extract that as separate commit.